packages/amplify-migration-codegen-e2e/src/helpers.ts

```diff
@@ -3,21 +3,164 @@ import {
   deleteProjectDir,
   initJSProjectWithProfile,
   addAuthWithDefault,
-  amplifyPushAuth,
+  amplifyPush,
   getProjectMeta,
   getUserPool,
   npmInstall,
   getNpxPath,
   nspawn as spawn,
+  addS3WithGuestAccess,
+  checkIfBucketExists,
+  addFunction,
+  functionBuild,
+  getFunction,
+  addApiWithoutSchema,
+  updateApiSchema,
+  getAppSyncApi,
+  amplifyPushForce,
+  describeCloudFormationStack,
 } from '@aws-amplify/amplify-e2e-core';
 import * as fs from 'fs-extra';
 import { $TSAny } from '@aws-amplify/amplify-cli-core';
 import path from 'node:path';
 import { CloudControlClient, GetResourceCommand } from '@aws-sdk/client-cloudcontrol';
+import { AppSyncClient, GetDataSourceCommand } from '@aws-sdk/client-appsync';
 import { unset } from 'lodash';
+import { createHash } from 'crypto';
+import { userInfo } from 'os';
 
+type AppId = string;
+type ProjectName = string;
+type BranchName = string;
+type SandboxName = string;
+
+type BackendIdentifier =
+  | {
+      namespace: Readonly<AppId>;
+      name: Readonly<BranchName>;
+      type: Readonly<'branch'>;
+      hash?: Readonly<string>;
+    }
+  | {
+      namespace: Readonly<ProjectName>;
+      name: Readonly<SandboxName>;
+      type: Readonly<'sandbox'>;
+      hash?: Readonly<string>;
+    };
+
+const STACK_NAME_LENGTH_LIMIT = 128;
+const AMPLIFY_PREFIX = 'amplify';
+const HASH_LENGTH = 10;
+const NUM_DASHES = 4;
 const pushTimeoutMS = 1000 * 60 * 20; // 20 minutes;
 
+function toStackName(backendId: BackendIdentifier): string {
+  const hash = getHash(backendId);
+
+  // only take the first 50 chars here to make sure there is room in the stack name for the namespace as well
+  const name = sanitizeChars(backendId.name).slice(0, 50);
+
+  const namespaceMaxLength =
+    STACK_NAME_LENGTH_LIMIT - AMPLIFY_PREFIX.length - backendId.type.length - name.length - NUM_DASHES - HASH_LENGTH;
+
+  const namespace = sanitizeChars(backendId.namespace).slice(0, namespaceMaxLength - 1);
+
+  return ['amplify', namespace, name, backendId.type, hash].join('-');
+}
+
```

Comment:
The logic to get the gen2 stackName from the backendId is the same and used fro m the gen2 library. Importing gen2 library would require making modifications to the gen2 packages as it is setup for ES6 modules only. Not worth to make changes just for e2e tests.

Reply:
Instead of replicating the code which is error prone and not future-proof, you can pipe the output of `npx ampx amplify --once` to a tmp file, there is a line ` Stack: 	<stackname>` in there.

Reply:
or you can search for `arn:aws:cloudformation:<region>:<accountid>:stack` prefix, it is a stack ARN . it is better since `Stack: <stackname>` may get changed

Reply:
another alternative is using `amplify sandbox --identifier <someId>`  then cfn.listStack() and substring match the stackname on the identifier

Reply:
Addressed. https://github.com/aws-amplify/amplify-cli/pull/13952/commits/d310c32e939e79bfc8ab2ae81be8d17123a4072f
I have piped the output of `npx ampx sandbox` and searched for the prefix `arn:aws:cloudformation:<region>:<accountid>:stack` having stack name.